### ISSUE TITLE
API to retrieve a docked widget from the viewer

### ISSUE THREAD

**Issue by haesleinhuepf** [napari/napari#3944]
## 🚀 Feature
I'd like to ask for an API extension of the viewer window: I would be happy to a) return a list of currently docked widgets and/or b) allow determining if specific widgets are docked already:

a)
```
for my_widget in viewer.window.get_dock_widgets(MyWidgetClass):
    # ...
```

b)
```
if viewer.window.is_widget_docked(my_widget):
   # ...
```

## Motivation

I casually access `Viewer._dock_widgets` to figure out if specific widgets are already docked to the viewer, before I dock another one. E.g. in [this code in napari-skimage-regionprops](https://github.com/haesleinhuepf/napari-skimage-regionprops/blob/master/napari_skimage_regionprops/_table.py#L161) I test if there is already a table showing the properties of a given layer. 

In another use-case: if a certain plugin-widget detects that other "sister" plugin-widgets are already activated / docked, it could grab them and tabbify itself with its family.

The current API offers:
* [`add_dock_widget`](https://github.com/napari/napari/blob/5d95bd4405f0997cf84153311a3aa7c46c3a2f3f/napari/_qt/qt_main_window.py#L715)
* [`remove_dock_widget`](https://github.com/napari/napari/blob/5d95bd4405f0997cf84153311a3aa7c46c3a2f3f/napari/_qt/qt_main_window.py#L876)

## Pitch

Add for example these two functions to `Viewer.window` (speculative pseudo-code):

```
def get_dock_widgets(self, widget_class):
    return [widget for widget in list(self._dock_widgets.values()) if isinstance(widget, widget_class)]

def is_widget_docked(self, widget):
    return widget in self._dock_widgets.values()
```

## Alternatives

Having both suggested methods above would give me some freedom. However, I think I could implement the desired functionality if I had either one of both.

## Additional context

I'm openin this feature request because my command line told me so:
```
...\napari-skimage-regionprops\napari_skimage_regionprops\_table.py:161: 
FutureWarning: Private attribute access ('Window._dock_widgets') in this context
(e.g. inside a plugin widget or dock widget) is deprecated and will be unavailable in version 0.5.0!
```
Hence, it's not urgent.

Furthermore, I would like to explicitly mention that this feature would be cool to enable communication between widgets of different plugins / vendors.

Thanks!

**Comment by tlambert03** [comment]
yep! definitely need something like this.  Want to give a little thought to exactly what parameters such as `get_widget` method would take.  but fully agree with the need for it.

**Comment by JoOkuma** [comment]
I really like the idea, I think that querying using the class could be better than using the string with the plugin name and class. 
However, I can think of some cases where the user would prefer to `type(widget) is Widget` rather than `isinstance`.

**Comment by tlambert03** [comment]
also, querying using class will only work for custom QWidget subclasses... not magicgui's (which are all `FunctionGui` instances)

**Comment by psobolewskiPhD** [comment]
Linking related/similar issue: https://github.com/napari/napari/issues/2203

**Comment by brisvag** [comment]
Closed in https://github.com/napari/napari/pull/7965!
